docs/jep/11759-python-package-structure.md

```diff
@@ -0,0 +1,142 @@
+    wheels would be built for the product of {CUDA/ROCm versions} x {OSes}
+    (e.g., `jaxlib-cuda-11-1-0.3.14-cp3-abi3-manylinux2014_x86_64.whl`).
+
+    We require that the versions of `jaxlib` and `jaxlib-cuda-...` match
+    exactly.
+
+    These wheels would also be structured to avoid a minor Python version
+    dependency, i.e., the `cuda` should work for all Python 3 versions we
+    support, rather than needing to be, say, Python 3.7 specific. This means
+    that we can only target the Python stable C API and must avoid the use of
+    `pybind11`.
+
+    It is to be determined whether we need only a single `jaxlib-cuda` plugin or
+    if we need multiple CUDA releases, however that is largely orthogonal to
+    this proposal.
+
+*   `jaxlib-tpu`: wheels containing TPU support as an optional plugin for the
+    `jaxlib` wheel. This wheel would be built only for Linux Cloud TPU VMs
+    (e.g., `jaxlib-tpu-0.3.14-cp3-abi3-linux_x86_64.whl`).
+
+
+## Why do this?
+
+Pros/cons of doing this:
+
+*   Simpler for users: users would be able to install a complete `jax` on CPU
+    with `pip install jax`. Provided we are able to get a size exception for our
+    GPU wheels, GPU wheels could be uploaded to Pypi and a GPU jax installation
+    could be installed with `pip install jax jaxlib-cuda-11-1`, without the need
+    for
+    [local versioning schemes](https://github.com/google/jax#pip-installation)
+    like `jaxlib+cuda110`that are not compatible with Pypi distribution.
+*   Reduced build times: `jaxlib` takes a long time to build; for example it
+    requires building XLA and LLVM. At present we cannot share build work
+    between, say, CUDA or Python versions, so to make a `jaxlib` release we
+    build ~12 different copies of XLA on Linux alone. By refactoring the
+    `jaxlib` wheels, we would remove the need to build the CPU support
+    repeatedly for each CUDA version, or the need to rebuild the GPU compiler
+    for each Python version.
+*   Splitting the wheels will reduce the total size of the binary distribution.
+    The largest `jaxlib` wheels at present are the CUDA wheels (~150MB each);
+    after refactoring we would no longer need to build one CUDA wheel per Python
+    version. This would reduce the size of our GPU wheel distribution by a
+    factor of 4, and it would make it easier to ship GPU wheels on Pypi.
+*   Making the compiler/runtime layer of JAX pluggable makes it easier to add
+    support for new compilers (IREE) and new hardware. In particular, we can
+    support new hardware from other vendors as new plugin packages without
+    causing a further combinatorial explosion in the number of builds.
+
+## Notes
+
+*   Hardware plugins like `jaxlib-cuda` will be distributed as Python packages
+    that act as plugins for `jaxlib`. There are standard Python mechanisms for
+    [creating and discovering plugins](https://packaging.python.org/guides/creating-and-discovering-plugins/)
```

Comment:
To my knowledge, the mechanisms mentioned here are for discovering plugins _that are installed_, not for discovering plugins _that exist_.  How would we notify a user who lacks such backend plugin(s) that their jaxlib will be hampered unless they further install something else (even on CPU, since apparently there would be a jaxlib-cpu optional backend as well, from what I'm understanding)?  If my understanding of all that is accurate, then doesn't this somewhat go against the opening premise that this is seeking to be
> Simpler for users: users would be able to install a complete `jax` on CPU with `pip install jax`.
?

Reply:
Correct, this is for discovering the plugins you have installed.

`jaxlib-cpu` we should require as a hard dependency, at least for now.

For a CUDA build, we can do two things:
a) We can suggest the user install `jax[cuda]` that adds `jaxlib-cuda` as a dependency.
b) We could also add code to core `jax` that detects if the user has `/dev/nvidia` devices but no `jaxlib-cuda` and suggests installing `jaxlib-cuda` (and similarly for other hardware for which there exists a plugin).

Reply:
> `jaxlib-cpu` we should require as a hard dependency, at least for now.

I wondered if that's how it would be done; I just didn't see that in the doc.

> `jax[cuda]`

Extras are a helpful strategy here, I agree.  (One possible gotcha is that the extra is on jax instead of on jaxlib, yet the == version match required from jaxlib to the extra and not from jax to the extra, which might get tricky.)

> We could also add code to core jax that detects if the user has /dev/nvidia devices but no jaxlib-cuda and suggests installing jaxlib-cuda (and similarly for other hardware for which there exists a plugin).

That would be helpful I think.  Please be aware that `/dev/nvidia*` device nodes, such as `/dev/nvidiactl`, are specific to Linux and to our discrete GPU drivers (Tegra drivers, for example, have `/dev/nvgpu` instead, and WSL2 has `/dev/dxg`; native Windows is another thing entirely).